src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -69,179 +69,188 @@ public Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object> arguments
     /// This will let one to hold onto <see cref="RemoteHostClient.Connection"/> for a while.
     /// this helper will let you not care about remote host being gone while you hold onto the connection if that ever happen
     /// 
-    /// and also make sure state is correct even if multiple threads call TryInvokeAsync at the same time. but this 
-    /// is not optimized to handle highly concurrent usage. if highly concurrent usage is required, either using
-    /// <see cref="RemoteHostClient.Connection"/> direclty or using <see cref="SessionWithSolution"/> would be better choice
+    /// when this is used, solution must be explicitly passed around between client (VS) and remote host (OOP)
     /// </summary>
     internal sealed class KeepAliveSession
     {
-        private readonly SemaphoreSlim _gate;
+        private readonly object _gate;
         private readonly IRemoteHostClientService _remoteHostClientService;
 
         private readonly string _serviceName;
         private readonly object _callbackTarget;
 
         private RemoteHostClient _client;
-        private RemoteHostClient.Connection _connection;
+        private ReferenceCountedDisposable<RemoteHostClient.Connection> _connectionDoNotAccessDirectly;
 
         public KeepAliveSession(RemoteHostClient client, RemoteHostClient.Connection connection, string serviceName, object callbackTarget)
         {
-            Initialize_NoLock(client, connection);
+            _gate = new object();
 
-            _gate = new SemaphoreSlim(initialCount: 1);
-            _remoteHostClientService = client.Workspace.Services.GetService<IRemoteHostClientService>();
+            Initialize(client, connection);
 
+            _remoteHostClientService = client.Workspace.Services.GetService<IRemoteHostClientService>();
             _serviceName = serviceName;
             _callbackTarget = callbackTarget;
         }
 
         public void Shutdown(CancellationToken cancellationToken)
         {
-            using (_gate.DisposableWait(cancellationToken))
+            ReferenceCountedDisposable<RemoteHostClient.Connection> connection;
+
+            lock (_gate)
             {
                 if (_client != null)
                 {
                     _client.StatusChanged -= OnStatusChanged;
                 }
 
-                _connection?.Dispose();
+                connection = _connectionDoNotAccessDirectly;
 
                 _client = null;
-                _connection = null;
+                _connectionDoNotAccessDirectly = null;
```

Comment:
:bulb: There is no advantage to setting `_connectionDoNotAccessDirectly` to null. It just complicates this method.

1. Calling dispose on the object more than once (even concurrently) will not put it in a bad state
2. It will not hold a strong reference to the underlying resource after dispose is called
3. It's possible to make the field `readonly` if we don't do this here